miprometheus/problems/video_text_to_class/cog/cog_utils/generate_dataset.py

```diff
@@ -0,0 +1,281 @@
+          if i % 10000 == 0 and i > 0:
+            print("Generated ", i, " examples")
+          example, _, _ = generate_example(max_memory, max_distractors,
+                                           family, epochs)
+          # Write the example to file
+          dump_str = json.dumps(example, sort_keys=True, separators=(',', ': '))
+          assert '\n' not in dump_str, 'dumps_str has new line %s' % (dump_str,)
+          f.write(dump_str.encode())
+          if i != examples_per_family - 1:
+            f.write(b'\n')
+
+  print("Wrote dataset into:", file_names)
+
+
+def log_exceptions(func):
+  @functools.wraps(func)
+  def wrapped_func(*args, **kwargs):
+    try:
+      return func(*args, **kwargs)
+    except BaseException as e:
+      print('Exception in ' + func.__name__)
+      traceback.print_exc()
+      raise e
+  return wrapped_func
+
+
+@log_exceptions
+def generate_train(output_dir, examples_per_family, start_index, per_file):
+  generate_dataset(FLAGS.epochs, FLAGS.max_distractors, FLAGS.max_memory,
+                   examples_per_family, output_dir, random_families=True,
+                   start_index=start_index, per_file=per_file,
+                   compress=FLAGS.compress)
+
+
+@log_exceptions
+def generate_val_or_test(data_type, cog_variant):
+  """
+  Args:
+    data_type: 'val' or 'test'
+  """
+  # 20x smaller than training.
+  output_dir = os.path.join(
+      FLAGS.output_dir, '%s_%s' % (data_type, cog_variant))
+  generate_dataset(FLAGS.epochs, FLAGS.max_distractors, FLAGS.max_memory,
+                   max(FLAGS.examples_per_family // 20, 50),
+                   output_dir, random_families=False,
+                   compress=FLAGS.compress)
+
+
+def main(argv):
+  cog_variant = '%d_%d_%d' % (FLAGS.epochs, FLAGS.max_memory, FLAGS.max_distractors)
+
+  train_output_dir = os.path.join(FLAGS.output_dir, 'train_' + cog_variant)
+
+  if FLAGS.parallel:
+    assert FLAGS.parallel > 2
+    train_parallel = FLAGS.parallel - 2
+    assert (FLAGS.examples_per_family % (2 * train_parallel)) == 0, (
+      "examples_per_family must be a multiple of 2*(parallel - 2)")
+    examples_per_family_per_job = FLAGS.examples_per_family // train_parallel
+
+    pool = multiprocessing.Pool(processes=FLAGS.parallel)
+    jobs = []
+    n_tasks = len(task_bank.task_family_dict)
+    for i in range_fn(train_parallel):
+      jobs.append(pool.apply_async(generate_train, (
+        train_output_dir, examples_per_family_per_job,
+        2 * i, n_tasks * examples_per_family_per_job // 2)))
+    jobs.append(pool.apply_async(generate_val_or_test, ('val', cog_variant)))
+    jobs.append(pool.apply_async(generate_val_or_test, ('test', cog_variant)))
+
+    # wait for all jobs to complete
+    [j.get() for j in jobs]
+  else:
+    generate_train(train_output_dir)
+    generate_val_or_test('val', cog_variant)
+    generate_val_or_test('test', cog_variant)
+
+
+
+if __name__ == '__main__':
+  tf.app.run(main)
```

Comment:
Ok, so this seems like a self-sustainable application that generates training/validation/testing sets.

From my point of view this should be:
a) incorporated in a problem in such a way that the problem will generate the dataset when it is not present.
b) on the other hand, it can also become a separate helper.

But what I am really planning is to "extract" from all problems pieces of code that generate/download datasets and implement a standalone helper (issue #79 ) that the user could use to download/generate them by instatiating problem and calling that method - it will be useful in case of all big datasets, when we do not want to run trainer just to learn that right now you have to wait for n hours...

Reply:
My plan was to incorporate the generation into the yaml configuration, such that a user can provide generation parameters in the config file and the dataset will be generated automatically. This would be best as part of a helper you described.